File: app/src/main/java/com/haoyu/app/fragment/PageProgressFragment.java
package com.haoyu.app.fragment;

import android.os.Bundle;
import android.text.Html;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.haoyu.app.base.BaseFragment;
import com.haoyu.app.entity.CourseProgressEntity;
import com.haoyu.app.entity.CourseProgressResult;
import com.haoyu.app.lego.student.R;
import com.haoyu.app.utils.Constants;
import com.haoyu.app.utils.OkHttpClientManager;
import com.haoyu.app.view.ColorArcProgressBar;
import com.haoyu.app.view.LoadFailView;
import com.haoyu.app.view.LoadingView;
import com.haoyu.app.view.RoundRectProgressBar;

import butterknife.BindView;
import butterknife.BindViews;
import okhttp3.Request;

/**
 * 创建日期：2016/11/30 on 9:49
 * 描述:  进度
 * 作者:马飞奔 Administrator
 */
public class PageProgressFragment extends BaseFragment {
    @BindView(R.id.loadingView)
    LoadingView loadingView;
    @BindView(R.id.loadFailView)
    LoadFailView loadFailView;
    @BindView(R.id.contentView)
    View contentView;
    @BindView(R.id.progressBar)
    ColorArcProgressBar progressBar;
    @BindViews({R.id.tv_time, R.id.tv_score, R.id.tv_state, R.id.tv_video, R.id.tv_html, R.id.tv_assignment, R.id.tv_test, R.id.tv_discussion, R.id.tv_survey})
    TextView[] textViews;
    @BindViews({R.id.ll_video, R.id.ll_html, R.id.ll_assignment, R.id.ll_test, R.id.ll_discussion, R.id.ll_survey})
    LinearLayout[] layoutViews;
    @BindViews({R.id.videoProgress, R.id.htmlProgress, R.id.assignmentProgress, R.id.testProgress, R.id.discussionProgress, R.id.surveyProgress})
    RoundRectProgressBar[] progressBars;
    private boolean training;
    private String courseId, courseType;
    private OnSelectCallBack onSelectCallBack;

    @Override
    public int createView() {
        return R.layout.fragment_page_progress;
    }

    @Override
    public void initView(View view) {
        Bundle bundle = getArguments();
        if (bundle != null) {
            training = bundle.getBoolean("training", false);
            courseId = bundle.getString("entityId");
            courseType = bundle.getString("courseType");
        }
        if (courseType != null && courseType.equals("微课")) {
            layoutViews[1].setVisibility(View.GONE);
            layoutViews[2].setVisibility(View.GONE);
            layoutViews[4].setVisibility(View.GONE);
            layoutViews[5].setVisibility(View.GONE);
        }
    }

    @Override
    public void initData() {
        String url = Constants.OUTRT_NET + "/" + courseId + "/study/m/course/" + courseId + "/study_progress";
        addSubscription(OkHttpClientManager.getAsyn(context, url, new OkHttpClientManager.ResultCallback<CourseProgressResult>() {
            @Override
            public void onBefore(Request request) {
                loadingView.setVisibility(View.VISIBLE);
            }

            @Override
            public void onError(Request request, Exception e) {
                loadingView.setVisibility(View.GONE);
                loadFailView.setVisibility(View.VISIBLE);
            }

            @Override
            public void onResponse(CourseProgressResult response) {
                loadingView.setVisibility(View.VISIBLE);
                if (response != null && response.getResponseData() != null) {
                    contentView.setVisibility(View.VISIBLE);
                    showData(response.getResponseData());
                }
            }
        }));
    }

    private void showData(CourseProgressEntity entity) {
        if (training) {
            if (entity.getmTimePeriod() != null && entity.getmTimePeriod().getMinutes() > 0) {
                textViews[0].setText(computeTimeDiff(entity.getmTimePeriod().getMinutes()));
            } else {
                if (entity.getmTimePeriod() != null && entity.getmTimePeriod().getState() != null) {
                    textViews[0].setText("课程" + entity.getmTimePeriod().getState());
                } else {
                    textViews[0].setText("课程进行中");
                }
            }
        } else {
            textViews[0].setText("课程已结束");
        }
        double score = entity.getScore();
        textViews[1].setText(String.valueOf((int) score));
        if (entity.getState() != null && entity.getState().equals("pass")) {
            textViews[2].setText("合格");
        } else if (entity.getState() != null && entity.getState().equals("nopass")) {
            textViews[2].setText("不合格");
        } else {
            textViews[2].setText("未登记");
        }
        progressBar.setMaxValues(100);
        progressBar.setCurrentValues((float) score);
        progressBars[0].setMax(entity.getActivityVideoNum());
        progressBars[0].setProgress(entity.getCompleteVideoNum());
        progressBars[1].setMax(entity.getActivityHtmlNum());
        progressBars[1].setProgress(entity.getCompleteHtmlNum());
        progressBars[2].setMax(entity.getActivityAssignmentNum());
        progressBars[2].setProgress(entity.getCompleteAssignmentNum());
        progressBars[3].setMax(entity.getActivityTestNum());
        progressBars[3].setProgress(entity.getCompleteTestNum());
        progressBars[4].setMax(entity.getActivityDiscussionNum());
        progressBars[4].setProgress(entity.getCompleteDiscussionNum());
        progressBars[5].setMax(entity.getActivitySurveyNum());
        progressBars[5].setProgress(entity.getCompleteSurveyNum());
        textViews[3].setText("已观看" + entity.getCompleteVideoNum() + "个/" + entity.getActivityVideoNum() + "个");
        textViews[4].setText("已观看" + entity.getCompleteHtmlNum() + "个/" + entity.getActivityHtmlNum() + "个");
        textViews[5].setText("已完成" + entity.getCompleteAssignmentNum() + "篇/" + entity.getActivityAssignmentNum() + "篇");
        textViews[6].setText("已完成" + entity.getCompleteTestNum() + "个/" + entity.getActivityTestNum() + "个");
        textViews[7].setText("已参与" + entity.getCompleteDiscussionNum() + "个/" + entity.getActivityDiscussionNum() + "个");
        textViews[8].setText("已完成" + entity.getCompleteSurveyNum() + "个/" + entity.getCompleteSurveyNum() + "个");
    }

    private CharSequence computeTimeDiff(long minutes) {
        String timeStr = "距离课程结束还有";
        StringBuilder actionText = new StringBuilder();
        if (minutes <= 0) { //1分钟内 服务端的时间 可能和本地的有区别 所以小于0的 对于这个情况全部都显示刚刚
            timeStr += "1分钟";
        } else if (minutes < 60) { // 1小时内
            timeStr += minutes + "分钟";
        } else if (minutes < 24 * 60) { // 一天内
            timeStr += (minutes / 60 == 0 ? 1 : minutes / 60) + "小时";
        } else if (minutes < 30 * 24 * 60) { // 天前
            long day = minutes / 60 / 24;
            long hour = (minutes - day * 24 * 60) / 60;
            long min = minutes - (day * 24 * 60) - (hour * 60);
            // 输出结果
            actionText.append("距离课程结束还有");
            actionText.append("<font color='#FEE42D'>"
                    + day + "</font>天");
            actionText.append("<font color='#FEE42D'>"
                    + hour + "</font>时");
            actionText.append("<font color='#FEE42D'>"
                    + min + "</font>分");
            timeStr = actionText.toString();
            return Html.fromHtml(timeStr);
        } else if (minutes < 12 * 30 * 24 * 60) { // 月
            timeStr += (minutes / 30 * 24 * 60 == 0 ? 1 : minutes / (30 * 24 * 60)) + "个月";
        } else {
            timeStr += (minutes / 12 * 30 * 24 * 60 == 0 ? 1 : minutes / (12 * 30 * 24 * 60)) + "年";
        }
        return timeStr;
    }

    @Override
    public void setListener() {
        loadFailView.setOnRetryListener(new LoadFailView.OnRetryListener() {
            @Override
            public void onRetry(View v) {
                initData();
            }
        });
        View.OnClickListener listener = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (onSelectCallBack != null)
                    onSelectCallBack.onClickCallBack();
            }
        };
        for (LinearLayout layout : layoutViews) {
            layout.setOnClickListener(listener);
        }
    }

    public void setOnSelectCallBack(OnSelectCallBack onSelectCallBack) {
        this.onSelectCallBack = onSelectCallBack;
    }

    public interface OnSelectCallBack {
        void onClickCallBack();
    }
}

File: app/src/main/java/com/haoyu/app/fragment/PageProgressFragment.kt
package com.haoyu.app.fragment

import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import com.haoyu.app.base.BaseFragment
import com.haoyu.app.entity.CourseProgressEntity
import com.haoyu.app.entity.CourseProgressResult
import com.haoyu.app.lego.student.R
import com.haoyu.app.utils.Constants
import com.haoyu.app.utils.OkHttpClientManager
import com.haoyu.app.utils.TimeUtil.computeTimeDiff
import com.haoyu.app.view.CircleProgressBar
import com.haoyu.app.view.LoadFailView
import com.haoyu.app.view.LoadingView
import com.haoyu.app.view.RoundRectProgressBar
import okhttp3.Request

/**
 * 创建日期：2018/1/15.
 * 描述:课程学习进度
 * 作者:xiaoma
 */
class PageProgressFragment : BaseFragment() {
    private lateinit var loadingView: LoadingView
    private lateinit var loadFailView: LoadFailView
    private lateinit var llContent: LinearLayout
    private lateinit var progressBar: CircleProgressBar
    private val textViews = arrayOfNulls<TextView>(10)
    private val layoutViews = arrayOfNulls<LinearLayout>(6)
    private val progressBars = arrayOfNulls<RoundRectProgressBar>(6)
    private var training = false
    private var courseId: String? = null
    private var courseType: String? = null
    private var onSelectCallBack: OnSelectCallBack? = null

    override fun createView(): Int {
        return R.layout.fragment_page_progress
    }

    override fun initView(view: View) {
        arguments?.let {
            training = it.getBoolean("training", false)
            courseId = it.getString("entityId")
            courseType = it.getString("courseType")
        }
        loadingView = view.findViewById(R.id.loadingView)
        loadFailView = view.findViewById(R.id.loadFailView)
        llContent = view.findViewById(R.id.ll_content)
        progressBar = view.findViewById(R.id.progressBar)
        textViews[0] = view.findViewById(R.id.tv_time)
        textViews[1] = view.findViewById(R.id.tv_score)
        textViews[2] = view.findViewById(R.id.tv_text)
        textViews[3] = view.findViewById(R.id.tv_state)
        textViews[4] = view.findViewById(R.id.tv_video)
        textViews[5] = view.findViewById(R.id.tv_html)
        textViews[6] = view.findViewById(R.id.tv_assignment)
        textViews[7] = view.findViewById(R.id.tv_test)
        textViews[8] = view.findViewById(R.id.tv_discussion)
        textViews[9] = view.findViewById(R.id.tv_survey)
        layoutViews[0] = view.findViewById(R.id.ll_video)
        layoutViews[1] = view.findViewById(R.id.ll_html)
        layoutViews[2] = view.findViewById(R.id.ll_assignment)
        layoutViews[3] = view.findViewById(R.id.ll_test)
        layoutViews[4] = view.findViewById(R.id.ll_discussion)
        layoutViews[5] = view.findViewById(R.id.ll_survey)
        progressBars[0] = view.findViewById(R.id.videoProgress)
        progressBars[1] = view.findViewById(R.id.htmlProgress)
        progressBars[2] = view.findViewById(R.id.assignmentProgress)
        progressBars[3] = view.findViewById(R.id.testProgress)
        progressBars[4] = view.findViewById(R.id.discussionProgress)
        progressBars[5] = view.findViewById(R.id.surveyProgress)
        if (courseType != null && courseType == "微课") {
            layoutViews[1]?.visibility = View.GONE
            layoutViews[2]?.visibility = View.GONE
            layoutViews[4]?.visibility = View.GONE
            layoutViews[5]?.visibility = View.GONE
        }
        for (layout in layoutViews) {
            layout?.setOnClickListener({
                onSelectCallBack?.onClickCallBack()
            })
        }
    }

    override fun initData() {
        val url = Constants.OUTRT_NET + "/" + courseId + "/study/m/course/" + courseId + "/study_progress"
        addSubscription(OkHttpClientManager.getAsyn(context, url, object : OkHttpClientManager.ResultCallback<CourseProgressResult>() {
            override fun onBefore(request: Request) {
                loadingView.visibility = View.VISIBLE
            }

            override fun onError(request: Request, e: Exception) {
                loadingView.visibility = View.GONE
                loadFailView.visibility = View.VISIBLE
            }

            override fun onResponse(response: CourseProgressResult?) {
                loadingView.visibility = View.VISIBLE
                if (response != null && response.responseData != null) {
                    llContent.visibility = View.VISIBLE
                    updateUI(response.responseData)
                } else {
                    loadFailView.visibility = View.VISIBLE
                }
            }
        }))
    }

    private fun updateUI(entity: CourseProgressEntity) {
        if (training) {
            if (entity.getmTimePeriod() != null && entity.getmTimePeriod().minutes > 0) {
                textViews[0]?.text = "距离课程结束还有：" + computeTimeDiff(entity.getmTimePeriod().minutes)
            } else {
                if (entity.getmTimePeriod() != null && entity.getmTimePeriod().state != null) textViews[0]?.text = "课程" + entity.getmTimePeriod().state else {
                    textViews[0]?.text = "课程进行中"
                }
            }
        } else {
            textViews[0]?.text = "课程已结束"
        }
        val score = entity.score
        textViews[1]?.text = score.toInt().toString()
        textViews[2]?.text = "课程得分"
        if (entity.state != null && entity.state == "pass") {
            textViews[3]?.text = "合格"
        } else if (entity.state != null && entity.state == "nopass") {
            textViews[3]?.text = "不合格"
        } else {
            textViews[3]?.text = "未登记"
        }
        progressBar.maxProgress = 100f
        progressBar.setProgress(score.toFloat(), true)
        progressBars[0]?.max = entity.activityVideoNum
        progressBars[0]?.progress = entity.completeVideoNum
        progressBars[1]?.max = entity.activityHtmlNum
        progressBars[1]?.progress = entity.completeHtmlNum
        progressBars[2]?.max = entity.activityAssignmentNum
        progressBars[2]?.progress = entity.completeAssignmentNum
        progressBars[3]?.max = entity.activityTestNum
        progressBars[3]?.progress = entity.completeTestNum
        progressBars[4]?.max = entity.activityDiscussionNum
        progressBars[4]?.progress = entity.completeDiscussionNum
        progressBars[5]?.max = entity.activitySurveyNum
        progressBars[5]?.progress = entity.completeSurveyNum
        textViews[4]?.text = "已观看" + entity.completeVideoNum + "个/" + entity.activityVideoNum + "个"
        textViews[5]?.text = "已观看" + entity.completeHtmlNum + "个/" + entity.activityHtmlNum + "个"
        textViews[6]?.text = "已完成" + entity.completeAssignmentNum + "篇/" + entity.activityAssignmentNum + "篇"
        textViews[7]?.text = "已完成" + entity.completeTestNum + "个/" + entity.activityTestNum + "个"
        textViews[8]?.text = "已参与" + entity.completeDiscussionNum + "个/" + entity.activityDiscussionNum + "个"
        textViews[9]?.text = "已完成" + entity.completeSurveyNum + "个/" + entity.completeSurveyNum + "个"
    }

    interface OnSelectCallBack {
        fun onClickCallBack()
    }

    fun setOnSelectCallBack(onSelectCallBack: OnSelectCallBack) {
        this.onSelectCallBack = onSelectCallBack
    }

}